### ISSUE TITLE
[feature request] Add matrix exponential

### ISSUE THREAD

**Issue by carlosgmartin** [pytorch/pytorch#9983]
Add matrix power as implemented by [numpy.linalg.matrix_power](https://docs.scipy.org/doc/numpy/reference/generated/numpy.linalg.matrix_power.html), matrix exponential as implemented by [scipy.linalg.expm](https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.linalg.expm.html), and matrix logarithm as implemented by [scipy.linalg.logm](https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.linalg.logm.html).

- [x] matrix_power
- [ ] matrix_exp
- [ ] matrix_log
- [ ] matrix_sqrt

**Comment by vishwakftw** [comment]
Are these worth porting into ATen? If so, I can take it up.

**Comment by zou3519** [comment]
Yes, please!

**Comment by vishwakftw** [comment]
@zou3519 @fmassa I went through the SciPy implementation for `matrixexp` (named `expm` in SciPy), and have a few questions:
- the implementation is added for sparse arrays as well. Should I be considering that, or is that supposed to be left for later?
- the implementation seems to take advantage of upper triangular matrices, and there are some optimized calculations done for the same. Should I take the same approach with `ATen` or is a general implementation sufficient?
Thank you.

**Comment by oleg-kachan** [comment]
Upvote for this, matrix exponential `expm`, logarithm `logm` and (fractional) matrix power `powm` which includes square root

**Comment by harpone** [comment]
The approach by @vadimkantorov seems to make most sense IMO because there's already a (differentiable) PyTorch function for SVD, and after that the exponentiation is just elemwise exp for the diagonal.

There's also an elegant (and probably useful) expression for the gradient. I don't actually know how the pytorch gradients are defined (yet), but I could probably dig up/work out the math for the gradient if there's interest for that...

**Comment by SsnL** [comment]
Already implemented

**Comment by harpone** [comment]
OK great to hear that @SsnL! Can you point out the function (also for reference to anyone else that happens to stumble on this thread)? Can't find it in the docs...

**Comment by vishwakftw** [comment]
This issue is partially resolved: `matrix_power` is implemented, whereas `matrix_exp` is not yet done.

**Comment by fmassa** [comment]
@harpone here it is for reference https://pytorch.org/docs/master/torch.html?highlight=matrix_power#torch.matrix_power
Note that it is only available in the nightly build, not in 0.4.1

**Comment by vadimkantorov** [comment]
@SsnL It would be nice if matrix_power also supported non-integer powers like 0.5 and other cases. It may require a full abs approach or some custom Newton-step approximation (saw it in some paper for a fast matrix square root estimation without doing an svd)

**Comment by SsnL** [comment]
@vadimkantorov Yep that would be nice. Do you still have a link to the paper?

**Comment by vadimkantorov** [comment]
@SsnL For Newton-step for matrix square root: https://arxiv.org/abs/1707.06772

**Comment by vishwakftw** [comment]
Since this involves a paper, maybe can this go into `contrib`?

**Comment by vadimkantorov** [comment]
I think a larger feature to decide is about supporting non-integer powers in matrix_power.

(The paper is just about a particular faster matrix square root approximation scheme - I agree, it can go to contrib.)

**Comment by carlosgmartin** [comment]
Should we add the matrix logarithm to this? With log and exp you can raise to fractional powers.

**Comment by harshs2** [comment]
Can developers please tell us that by when we should expect the implementations of expm and logm functions in pytorch?

**Comment by ferrine** [comment]
This might be helpful
https://github.com/SkafteNicki/cuda_expm

and CC @SkafteNicki

**Comment by ferrine** [comment]
That guy seems to have implemented this as well. I tried to build his module but gut stuck with linkage issues. The pure pytorch example was working correctly

**Comment by SkafteNicki** [comment]
@ferrine I never got the cuda implementation to work, because I got stuck with some linkage to cuSOLVER. The pure pytorch example works fine and can back-prop gradients. The cuda implementation was just to get a faster implementation, since I use matrix exponentials quite a bit myself.

**Comment by ferrine** [comment]
I've managed to implement (with a simple test) matrix exponential using torch script here:
https://github.com/ferrine/geoopt/blob/master/geoopt/linalg/_expm.py
I can open a PR to add that to pytorch master as well

**Comment by t-vi** [comment]
Just in case you need a taste for the limitations of various matrix exponential algorithms: https://www.cs.cornell.edu/cv/ResearchPDF/19ways+.pdf

**Comment by zhangguanheng66** [comment]
Seems that the feature request has been completed. Feel free to re-open it if you have any questions.

**Comment by SsnL** [comment]
hmm i don’t think this exists in pytorch tho. @zhangguanheng66

**Comment by zhangguanheng66** [comment]
@SsnL I thought you already implemented, as mentioned above. Anyways, I will re-open it and triage.

**Comment by calincru** [comment]
Any updates on the development of `expm`, `logm`, `sqrtm`?

**Comment by Lezcano** [comment]
The formula for the gradient of the exponential is derived in Proposition 6.1 here:
https://arxiv.org/abs/1909.09501
and an implementation for 32-bit skew-symmetric matrices can be found here: 
https://github.com/Lezcano/expRNN/blob/e29e9ec4e1d419763f2edbdd983f479177771f05/trivializations.py#L10

The following things would be missing, which should be very easy to implement modulo making some design choices:
- Implement the exponential for general matrices following Proposition 6.1 in the paper.
- Generalise the implementation to correctly handle batches of matrices of the same size.

Design choices:
- Choose how to compute the inverse needed in the gradient formula in general, as one of the most compelling use cases of the exponential is when its argument is skew-symmetric, and in this case computing the inverse is just computing the transpose. It would not be a good design to choose to compute the inverse, as it is much more costly and it may incur in unnecessary precision errors.
- Leave the 32 bit implementation, or implement both a 32 and 64 bit implementation, or just the 64 bit implementation?
- Just use a pade7 (or pade13 for 64-bits) when handling batches of matrices (it might incur in a loss of precision when the elements in the batch have small entries) or split the batch into the subbatches of matrices that need the different approximations and then put them together again (more computationally costly).

If someone tells me which of these implementation decisions follow closer the current Pytorch guidelines, I'm happy to implement them myself and PR them

**Comment by soumith** [comment]
@vishwakftw could you guide @Lezcano

**Comment by vishwakftw** [comment]
@soumith I'd be happy to.

@Lezcano perhaps you could try implementing the matrix exponential first, and then we can follow up with the gradient implementation.

Regarding your design choices: it would be preferable to have separate implementations based on precision. IIRC, we would have to make one more approximation for the 64-bit case.

**Comment by Lezcano** [comment]
I put together an implementation of the exponential of matrices and its gradients for arbitrary matrices, which chooses between the 32 and the 64 bit algorithm, and does some optimization when the input is a skew symmetric matrix. It also implements some (very) minimal tests 
https://github.com/Lezcano/expm
It does not implement `expm` for batches of matrices. This could be done with a bit of effort, if one wants to go for precision (choosing the best Pade approximant for each matrix in the batch) or it should be almost straightforward if one chooses to take the maximum degree Pade approximant for the whole batch.

**Comment by vishwakftw** [comment]
I think we should go with the maximum degree Pade approximation for all matrices in the batch.

@Lezcano we have our forward implementations in C++ residing in aten/src/ATen/native/LinearAlgebra.cpp (or a separate file MatrixFunctions.cpp) The derivatives go in tools/autograd/templates/Functions.cpp.

Would it be possible for you to port your code to C++

**Comment by Lezcano** [comment]
I could do that, but I am afraid that I don't have the time now. I will have some time for this again in December, but I will be quite busy until then.
I will leave the implementation there for now (it should be trivial to extend to batches in Pytorch) in case someone wants to go ahead and port the code. If not, I will do it in December.

**Comment by vishwakftw** [comment]
@Lezcano thanks. If I find time, I'll try doing it myself, or I'll leave it to anyone else to do it too. Otherwise, we'll wait for you.

Feel free to reach out with any questions you might have about the implementation anytime.

**Comment by zasdfgbnm** [comment]
Instead of hunting for the "best" implementation and keep this feature being delayed, is there anything preventing us from adding a naive implementation first?  Something like:
```python
def matrix_exp(matrix, symmetric=False):
    if not symmetric:
        raise NotImplementedError('matrix exp for non-symmetric matrix is not implemented')
    e, V = torch.symeig(matrix, eigenvectors=True)
    return V.t() @ torch.diag(e).exp() @ V
```

In general, all the matrix functions, as long as this function could be written as a power series, could be implemented using the above method
```python
def matrix_func(matrix, func, symmetric=False):
    if not symmetric:
        raise NotImplementedError('matrix exp for non-symmetric matrix is not implemented')
    e, V = torch.symeig(matrix, eigenvectors=True)
    return V.t() @ func(torch.diag(e)) @ V
```

We can always improve later.

**Comment by ferrine** [comment]
BTW, I have a working jit scripted version of this if it is relevant
https://github.com/geoopt/geoopt/blob/master/geoopt/linalg/_expm.py

**Comment by Lezcano** [comment]
@ferrine Your implementation of the forward pass is correct, as I see that is an adaptation of numpy, as it is mine, but it's missing the implementation of an approximation for the gradient.
@zasdfgnm an implementation using the eigendecomposition, not only just works for symmetric matrices, but the algorithm is rather poor in terms of numerical stability.

A stronger way to go forward would be taking the implementation in
https://github.com/Lezcano/expm
and translating it to C++. As I said, I might find time to do this during Christmas, but if anyone wants to chip in before that, any help is welcome.
 
If someone wants to go the extra mile, this paper gives an algorithm that's purely polynomial based, and should by faster in GPUs:
http://personales.upv.es/serblaza/2018Expm_Rev.pdf
To approximate the gradient, I implemented a simple algorithm that works for generic functions, but a more efficient algorithm is known and detailed in:
https://pdfs.semanticscholar.org/e537/b375e59ca61ee5243c0dfeeefa9329f63725.pdf
This algorithm is implemented in scipy as well, under the function `scipy.linalg.expm_frechet`, and the method SPS
https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.expm_frechet.html?highlight=frechet

**Comment by chrisby** [comment]
Any updates on this?

**Comment by Lezcano** [comment]
I just updated today the implementation in the repo to handle batches of matrices. It would be a matter of porting it to ATen and writing the accompanying testing. Alas, I do not have the time at the moment to do so, but maybe someone could chip in? Most of the work is already done, as the algorithms are already there and I have tested them.

**Comment by tellihichem** [comment]
:red_circle: I am very disappointed on ~~_pytorch_~~ :confused:  for not implementing matrix log (logm in matlab), which is a very essential tool in machine learning and tensor manipulation. I am thinking to switch back to matlab.

**Comment by zou3519** [comment]
I'm bumping the priority on this feature due to the amount of activity in this thread

**Comment by nikitaved** [comment]
Is anyone working on it? I could give it a try.

**Comment by Lezcano** [comment]
Not that I know. The state of affairs is that we have an implementation that works for batches and all in 
https://github.com/Lezcano/expm
(in particular, the Taylor one, as it is faster in GPU), and that should be ported to ATen.

The implementation currently uses a poor man's version of `torch.matrix_power` that accepts a sequence of indices in the exponent term. A good implementation of the matrix exponential would go through extending first `torch.matrix_power` to handle a vector (or a tensor) in its second argument. Feel free to pop me a message if you have any questions about the algorithm.

**Comment by mdanb** [comment]
Any updates on this?

**Comment by nikitaved** [comment]
@mdanb , sorry, been busy. I want to finish something before jumping on this one. It is right next on my bucket list.

**Comment by Lezcano** [comment]
For what is worth, regarding the `matrix_sqrt` and `matrix_log`, both implementations rely on the real Schur form, as implemented in Scipy, so that should be implemented in PyTorch. As all these functions are analytic, the same algorithm used to compute the gradient for the `expm` as implemented in the repo above would work for these two (see lines 263 and 286 in https://github.com/Lezcano/expm/blob/master/pytorch_expm/expm_taylor.py)

**Comment by nikitaved** [comment]
@Lezcano , thank you! You recommend the method from http://personales.upv.es/serblaza/2018Expm_Rev.pdf , right? Any recommendations on the backward step?

**Comment by Lezcano** [comment]
The backward step can be implemented the same way for all these functions, as they are all analytic, the algorithm employed is the same. It is based on this paper:
https://dl.acm.org/doi/10.1137/S0895479895283409
and it's implemented in pure PyTorch for a generic analytic function here:
https://github.com/Lezcano/expm/blob/master/pytorch_expm/expm_taylor.py#L263

About the forward method for the `expm`, yes, that's the paper I think that would works best for GPUs. I implemented their algorithm here, implementing also an exact backward step (the one mentioned before using Mathias' algorithm):
https://github.com/Lezcano/expm/blob/master/pytorch_expm/expm_taylor.py
I already implemented it as fast as I could, but maybe you find some other optimizations, probably the function `matrix_power_two_batch` is the one that could be optimized the most (this is just a batched version of `matrix_power`, so it would be best to just implement this in `torch.matrix_power`).

**Comment by nikitaved** [comment]
@Lezcano , correct me if I am wrong, but it seems to me the paper decides the parameters based on the matrix 1-induced norm, not a vector 1-norm? It looks like in your code you use  l1 norm for a single matrix, and Frobenius norm for batches...

**Comment by nikitaved** [comment]
And, by the way, does PyTorch support induced norms? Would probably not hurt to implement at least 1-induced for this case I guess.

**Comment by Lezcano** [comment]
You are completely right! That frobenius norm was left there, as some previous version of torch had a bug when selecting a `p =/= 2` and selecting some custom `dim`. And yes, it should really be the `1`-operator norm according to the paper. It would be good to have this on Pytorch anyway, so it might be good to add it to `torch.norm`? At least the simple ones, like operator norms 1, 2, and \infty.

**Comment by nikitaved** [comment]
It seems to me the paper has a typo in `expm2`, there I would expect the condition `||A|| <= theta`.

Also, thinking about the precision it seems to me we can just go with `T_18`. The values of `theta` are pretty small fort `T_i, i < 18` to be realizable in practice. Why? I would assume that in the context of neural networks most matrices are coming from the Gaussian distribution (say, Xavier initialization). The expected operator norm 1 has a lower bound of an expectation of max Gaussians, and in this case it is at least `Theta(sqrt(log n))`, where `n` is the size of a matrix.

So, I propose to stick to `T_18`, it is just much more likely, and make the implementation easier to understand/debug, and portable across CPU/CUDA (only high level uniform batch operations).

**Comment by Lezcano** [comment]
For batching, the current implementation is already done just with T_18 (and handling two trivial cases separately). At the moment, it is just for the case of one matrix when the other cases are used. 

On the other hand, although I do think that for some cases just T_18 is going to be useful, there are other cases where the other polynomials save quite a bit of time. I do not think that the Xavier initialization will be used without further constraints together with the exponential as, in the case of one variable, the exponential of matrices is fairly bad conditioned.

One example when it is not bad conditioned is when doing optimisation over the orthogonal matrices. In this case, one way to do this (and a quite efficient one) is to sample a random orthogonal matrix B, and optimising the function `A -> Bexpm(A-A.t())` with initial condition `A = 0`. In this case, if the norm of `A` remains small during some reasonable time (which is usually the case), cheaper approximations save quite a bit of time.

**Comment by nikitaved** [comment]
It is quite unlikely that Gassian random matrices have large eigenvalues, actually.

Anyway, I do believe it is imperative to bring the feature first, and then optimize as needed.
The efficient batched version will require device-specific code for sure, and it gets more involved for sure...

**Comment by Lezcano** [comment]
True, Gaussian random matrices do not have large eigenvalues on average, but if you backprop throu `expm`, the gradients will make them very large in no time, same as it happens with the regular exponential.

My use case is exactly the one I outlined above, but in more generality: Optimisation on manifolds. Optimisation with orthogonal constraints is just one instance of it.

And yes, you are right, maybe it is best to have a version that can be shipped, and then, from there, implement the other optimisations. I believe that this should not be too difficult after the first version is finished.

**Comment by nikitaved** [comment]
@Lezcano , could you please have a look at the mathematical part of my PR. It is ready for review.

**Comment by Lezcano** [comment]
I left a review in the PR---it's looking very good.
We would just be missing a way to implement the batches in parallel as, at the moment, that sequential operation will surely be a bottleneck for reasonable batches. I have also sent an email to the authors of the paper asking them about that numerical mismatch.

**Comment by nikitaved** [comment]
For more or less efficient batch processing we need native batch loop. It can be implemented in a separate PR.

**Comment by nikitaved** [comment]
Yeah, the CUDA batched version is quite slow. I will work on the device-specific implementation then.

**Comment by nikitaved** [comment]
When it comes to the matrix square root, would you be interested in the negative square root? I guess it is useful to have, especially for psd matrices, what do you think? It also makes the implementation a package deal if implemented via the matrix sign function.

**Comment by Lezcano** [comment]
I have not looked into the implementation of the square root and related functions, but they would surely be very useful! I think that, if it's implemented in terms of the real Schur decomposition would allow to directly implement arbitrary fractional powers, although this could be a separate algorithm.

As PyTorch does not currently support complex numbers, this would have to be a real Schur decomposition, for which some algorithms exist, although I don't know how fast are they on GPU. In any case, having an implementation of the real Schur decomposition would make "fairly straightforward" the implementation of both A^p where p is a rational number and the logarithm. Higham has several articles on these type of algorithms which are the (CPU) standard, cf.

    Nicholas J. Higham and Lijing lin (2011) “A Schur-Pade Algorithm for Fractional Powers of a Matrix.” SIAM Journal on Matrix Analysis and Applications, 32 (3). pp. 1056-1078. ISSN 0895-4798

    Awad H. Al-Mohy and Nicholas J. Higham (2012) “Improved Inverse Scaling and Squaring Algorithms for the Matrix Logarithm.” SIAM Journal on Scientific Computing, 34 (4). C152-C169. ISSN 1095-7197

This second one is "similar" to his article about using Padé approximants to approximate the matrix exponential.

**Comment by nikitaved** [comment]
Amazing, but all these things can be implemented with a sign function (logarithm, fractional power,...)! So, to bring things to the table fast, I will probably concentrate on this path.

**Comment by ezyang** [comment]
> As PyTorch does not currently support complex numbers

We support complex on master!

**Comment by nikitaved** [comment]
@ezyang , unfortunately, some of the fundamental functionality is missing yet :(
```python
In [1]: import torch

In [2]: a = torch.randn(3, 3, dtype=torch.complex128)

In [3]: torch.matmul(a, a)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-3-7a0abd112090> in <module>
----> 1 torch.matmul(a, a)

RuntimeError: _th_addmm_out not supported on CPUType for ComplexDouble

```

**Comment by ezyang** [comment]
Yeah, addmm has been blocked on getting ported to ATen first :( https://github.com/pytorch/pytorch/issues/24679

**Comment by nikitaved** [comment]
I will try to use algorithms which do not rely on the explicit complex structure, and the matrix sign function seems like a good solution. The complex support will be enabled automatically once `matmul` supports complex numbers.

**Comment by oumarkaba** [comment]
Are there any updates on this? I'm looking forward to the matrix log!

**Comment by mruberry** [comment]
Sorry, no update on matrix log or matrix sqrt. We are planning to review better SciPy linear algebra compatibility in the near future, however.

**Comment by rgommers** [comment]
@nikitaved has done work on matrix sqrt, which should be unblocked once his `linalg.lstsq` PR is in - it's still a big job after that though, so no ETA. No one has worked on matrix log AFAIK.

**Comment by annikabrundyn** [comment]
also looking forward to matrix sqrt! thanks for sharing an update on this

**Comment by nikitaved** [comment]
The issue with the matrix sqrt that I see is its implementation for CUDA, it is not so straightforward to translate the SciPy's implementation to the gpu (no Schur decomposition, dependence in loops). It is possible, however, to implement an iterative algorithm either only for the GPU, or for both the CPU and CUDA. If we agree on iterative algorithms, then I could start working on it.

**Comment by adizhol** [comment]
Any idea when matrix log will be available?

https://www.tensorflow.org/api_docs/python/tf/linalg/logm

**Comment by Lezcano** [comment]
As @nikitaved mentioned above, it is particularly tricky to implement `logm` on GPU for a number of reasons. We implement a poor man's version of `logm` mixing CPU and CUDA code for 1.11. Until then, given the extended compatibility between Numpy / Scipy and PyTorch, you can compute matrix logarithms of square matrices as follows:
```python
import scipy.linalg
import torch

def logm(A):
    assert A.dtype in (torch.float32, torch.float64, torch.complex64, torch.complex128)
    return torch.from_numpy(scipy.linalg.logm(A.cpu(), disp=False)[0]).to(A.device)
```

Note that if the tensor `A` is on GPU this will copy it to CPU, perform the computation on CPU, and copy it back to GPU. This is somewhat expensive, but gets the job done :)

**Comment by Lezcano** [comment]
Not much yet for the same reasons as for the `logm`: All the algorithms to approximate it are iterative, and iterative algorithms don't play well with GPUs. In the meantime, a similar trick as the one above with `logm` is the best we have:

```python
import scipy.linalg
import torch

def adjoint(A, E, f):
    A_H = A.T.conj().to(E.dtype)
    n = A.size(0)
    M = torch.zeros(2*n, 2*n, dtype=E.dtype, device=E.device)
    M[:n, :n] = A_H
    M[n:, n:] = A_H
    M[:n, n:] = E
    return f(M)[:n, n:].to(A.dtype)

def sqrtm_scipy(A):
    return torch.from_numpy(scipy.linalg.sqrtm(A.cpu(), disp=False)[0]).to(A.device)

class Sqrtm(torch.autograd.Function):
    @staticmethod
    def forward(ctx, A):
        assert A.ndim == 2 and A.size(0) == A.size(1)  # Square matrix
        assert A.dtype in (torch.float32, torch.float64, torch.complex64, torch.complex128)
        ctx.save_for_backward(A)
        return sqrtm_scipy(A)

    @staticmethod
    def backward(ctx, G):
        A, = ctx.saved_tensors
        return adjoint(A, G, sqrtm_scipy)

sqrtm = Sqrtm.apply

A = torch.rand(3, 3, dtype=torch.float64, requires_grad=True)
torch.autograd.gradcheck(sqrtm, A)
A = torch.rand(3, 3, dtype=torch.complex128, requires_grad=True)
torch.autograd.gradcheck(sqrtm, A)
```

**Comment by oleg-kachan** [comment]
@nikitaved, @Lezcano any progress with implementing `logm` into pytorch? How the issues you have mentioned are worked around in the tensorflow implementation https://www.tensorflow.org/api_docs/python/tf/linalg/logm?

**Comment by IvanYashchuk** [comment]
Those issues are not fixed in TF. TF doesn't provide any better functionality or substantial performance improvements than `scipy.linalg.logm`. Unfortunately, the documentation of `logm` in TF doesn't mention its limitations:

1. There is no derivative rule implemented
    https://github.com/tensorflow/tensorflow/issues/26321.
2. The implementation in TF is CPU-only using Eigen ([source code for the curious](https://github.com/tensorflow/tensorflow/blob/d929f08a7c075f9a88a138f2a1648d0e29c8ed41/tensorflow/core/kernels/linalg/matrix_logarithm_op.cc#L45)).

**Comment by jungerm2** [comment]
Here's a simple implementation of matrix logarithm that does not use scipy. Importantly, it can run on devices other than CPUs and be backpropagated through, but it assumes that the matrix is diagonalizable. Hopefully, this helps:

```python
def torch_logm(A):
    # https://en.wikipedia.org/wiki/Logarithm_of_a_matrix#Calculating_the_logarithm_of_a_diagonalizable_matrix
    lam, V = torch.linalg.eig(A)
    V_inv = torch.inverse(V).to(torch.complex128)

    A_prime = V_inv @ A.to(torch.complex128) @ V.to(torch.complex128)
    log_A_prime = A_prime * ~torch.eye(3, dtype=bool, device=A.device) + torch.diag(A_prime.diag().log())
    return V.to(torch.complex128) @ log_A_prime @ V_inv
```

**Comment by Lezcano** [comment]
The issue with this implementation is that the computation is *very* badly conditioned, as it doesn't exist a stable algorithm to compute the eigenvalues of an arbitrary matrix. 

This implementation in particular has a number of other issues. For example, it computes the inverse of the eigenvalues explicitly, which is also problematic. Even more, it recomputes the eigenvalues again and calls them `A_prime` for no very good reason.

If you want a stable and numerically correct implementation of this function, albeit with some limitations due to the use of SciPy, see https://github.com/pytorch/pytorch/issues/9983#issuecomment-891777620. Feel free to have a peak at the SciPy implementation for this function, to see why is it so tricky to implement, and even more, why is it so tricky to implement in a GPU (it has lots of data-dependent parts).

**Comment by jungerm2** [comment]
Thanks for the quick response! I've edited the above to remove the extra computation of `A_prime` (good catch). My use case is rotation matrices and the likes so the Eigen decomposition shouldn't be an issue I think. The main advantage for me is that this is roughly 20x faster than scipy on CPU. 

Also, why is computing the inverse of the eigenvectors explicitly a problem? How would you go about not doing so?

**Comment by Lezcano** [comment]
These questions related to classical linear algebra may be better suited for the forum: https://discuss.pytorch.org/ Feel free to ping me there and I may be able to help.

**Comment by husisy** [comment]
Hi, there. I implement an approximate version of matrix logarithm based on the discussion above (support GPU, batch, complex). It works well in my course work (density matrix, eigenvalue between 0 and 1, quantum stuff). you may try it if you need one.

https://github.com/husisy/matrix-logarithm

(I am not sure whether it works in general and how accurate it is. Also, its performance maybe not good)

**Comment by vadimkantorov** [comment]
For constant integer exponent matrix powers, it may also be implemented / or lowered as a product of power-of-two-exponent matrix powers (corresponding to binary representation of the exponent) https://github.com/pytorch/pytorch/pull/84569#issuecomment-1336497442

**Comment by lezcano** [comment]
That's currently available in `linalg.matrix_power`.

**Comment by vadimkantorov** [comment]
Then it's interesting whether it uses https://en.wikipedia.org/wiki/Exponentiation_by_squaring and when it makes sense for lowering...

**Comment by lezcano** [comment]
It is implemented this way, yes. Any other way would be just too slow.
https://github.com/pytorch/pytorch/blob/777ac632fb989119a95a95737d3d31c2ad1e52fe/aten/src/ATen/native/LinearAlgebra.cpp#L655-L667

**Comment by vadimkantorov** [comment]
Probably worth mentioning this in the docs explicitly. I also wonder if it's supported for various scripting / onnx export. At least, does it lower x.matrix_power(2) to x @ x?

**Comment by CompRhys** [comment]
Can I BUMP this with another request for torch.matrix_log?

**Comment by vadimkantorov** [comment]
@lezcano I also wonder if now iterative linalg algos can be better expressed with torch.while_loop / torch.cond

**Comment by lezcano** [comment]
those ops are designed for `torch.compile`. It is possible to implement these functions in triton (not without a fair amount of effort tho) and probably beat any existing implementation when it comes to speed.
The issue here is that we don't deploy triton functions in PyTorch core as triton does not have a good AoT story. But this would be 100% doable as a prototype in an external library, and we could consider porting it to core if/when PyTorch core figures out how to deploy triton kernels in eager mode.

**Comment by vadimkantorov** [comment]
I wonder if it's possible to add in triton functions in core ins some sort of torch.experimental namespace in core which is understood to be alpha/non-stable and for feedback collection without guarantees (and e.g. these functions can print always a warning to the screen)

Or can they be somehow only available in compile mode? Where triton is actually generated anyway

**Comment by lezcano** [comment]
If you want that, it'd be better to implement such a library externally to see the quality of those implementations and see if community uses them.

**Comment by yun520-1** [comment]
Adding `torch.linalg.matrix_power`, `torch.linalg.expm`, and `torch.linalg.logm` is a completeness gap. PyTorch already has `torch.linalg.matrix_norm` and `torch.linalg.det`; the missing piece is the exponential family. The implementation should dispatch to SciPy for the general case and use a fast path for small matrices to avoid the SciPy overhead.
